daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -591,39 +591,28 @@ convertExpr env0 e = do
         pure $
             ETmLam (varV1, TList TParty) $ ETmLam (varV2, TContractId t') $
             EUpdate $ UExercise tmpl' (mkChoiceName "Archive") (EVar varV2) (EVar varV1) mkEUnit
-    go env (VarIs "$dminternalFetchByKey") (LType t@(TypeCon tmpl []) : LType key : _dict : args)
-        = fmap (, args) $ do
-        key' <- convertType env key
-        tmpl' <- convertQualified env tmpl
-        let contractType = TConApp tmpl' []
-        let cidType = TContractId contractType
-        let fields = [("contractId", cidType), ("contract", contractType)]
-        (repackStruct, tupleType) <- mkRepackStruct env fields varV2
-        pure $ ETmLam (varV1, key') $
-          EUpdate $ UBind
-            (Binding
-              (varV2, TTuple fields)
-              (EUpdate (UFetchByKey RetrieveByKey
-                { retrieveByKeyTemplate = tmpl'
-                , retrieveByKeyKey = EVar varV1
-                })))
-            (EUpdate (UPure tupleType repackStruct))
-    go env (VarIs "$dminternalLookupByKey") (LType t@(TypeCon tmpl []) : LType key : _dict : args)
-        = fmap (, args) $ do
-        key' <- convertType env key
-        tmpl' <- convertQualified env tmpl
-        pure $ ETmLam (varV1, key') $
-          EUpdate $ ULookupByKey RetrieveByKey
-            { retrieveByKeyTemplate = tmpl'
-            , retrieveByKeyKey = EVar varV1
-            }
+    go env (VarIs f) (LType t@(TypeCon tmpl []) : LType key : _dict : args)
+        | f == "$dminternalFetchByKey" = conv UFetchByKey
+        | f == "$dminternalLookupByKey" = conv ULookupByKey
+        where
+            conv prim = fmap (, args) $ do
+                key' <- convertType env key
+                tmpl' <- convertQualified env tmpl
+                pure $ ETmLam (varV1, key') $
+                  EUpdate $ prim RetrieveByKey
+                    { retrieveByKeyTemplate = tmpl'
+                    , retrieveByKeyKey = EVar varV1
+                    }
     go env (VarIs "unpackPair") (LType (StrLitTy f1) : LType (StrLitTy f2) : LType t1 : LType t2 : args)
         = fmap (, args) $ do
             t1 <- convertType env t1
             t2 <- convertType env t2
             let fields = [(mkField f1, t1), (mkField f2, t2)]
-            (repackStruct, _) <- mkRepackStruct env fields varV1
-            pure $ ETmLam (varV1, TTuple fields) repackStruct
+            tupleTyCon <- qGHC_Tuple env (mkTypeCon ["Tuple" ++ show (length fields)])
```

Comment:
I kind of liked having the logic for this in a separate function. Maybe add something like `mkTupleType :: [(FieldName, LF.Type)] -> ConvertM LF.Type` that only contains the type logic of `mkRepackStruct`.

Reply:
In reality it looks less nice than I imagined. I'll leave it as is.